# HR Analytics - Employee Performance & Retention Analysis

## 1. Problem Statement

The objective of this project is to analyse HR data to assess employee performance, identify trends influencing retention, and provide actionable recommendations to improve employee productivity and reduce turnover.

Specifically, this project aims to:
- Identify performance patterns based on factors such as age, education, department, and training.
- Examine retention trends based on employee characteristics including tenure, performance ratings, and awards.
- Analyse relationships between key employee metrics to recommend strategies for improving employee engagement and retention.

## 2. Importing Libraries & Loading Dataset

In [21]:
import pandas as pd
df = pd.read_csv(r'D:\Projects\employees_dataset.csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17417 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17417 non-null  int64  
 1   department             17417 non-null  object 
 2   region                 17417 non-null  object 
 3   education              16646 non-null  object 
 4   gender                 17417 non-null  object 
 5   recruitment_channel    17417 non-null  object 
 6   no_of_trainings        17417 non-null  int64  
 7   age                    17417 non-null  int64  
 8   previous_year_rating   16054 non-null  float64
 9   length_of_service      17417 non-null  int64  
 10  KPIs_met_more_than_80  17417 non-null  int64  
 11  awards_won             17417 non-null  int64  
 12  avg_training_score     17417 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.7+ MB


In [22]:
df.head()

employee_id         department     region  education gender  \
0         8724         Technology  region_26  Bachelors      m   
1        74430                 HR   region_4  Bachelors      f   
2        72255  Sales & Marketing  region_13  Bachelors      m   
3        38562        Procurement   region_2  Bachelors      f   
4        64486            Finance  region_29  Bachelors      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   24                   NaN   
1               other                1   31                   3.0   
2               other                1   31                   1.0   
3               other                3   31                   2.0   
4            sourcing                1   30                   4.0   

   length_of_service  KPIs_met_more_than_80  awards_won  avg_training_score  
0                  1                      1           0                  77  
1                  5                      0           0                  51  
2                  4                      0           0                  47  
3                  9                      0           0                  65  
4                  7                      0           0                  61

In [23]:
df.shape

(17417, 13)

## 3. Dataset Overview

**Source:**  
The dataset is sourced from Internshala Trainings – HR Analytics Case Study and provided in CSV (.csv) format.

**Dataset Shape:**  
- **Records:** 17,417 employees  
- **Features:** 13 columns  

**Column Summary:**
- `employee_id`: Unique employee identifier
- `department`: Employee’s department
- `region`: Employee location/region code
- `education`: Education qualification
- `gender`: Employee gender
- `recruitment_channel`: Hiring source
- `no_of_trainings`: Trainings completed in one year
- `age`: Employee age
- `previous_year_rating`: Last year's performance rating (1–5)
- `length_of_service`: Work tenure in years
- `KPIs_met_more_than_80`: KPI performance above 80% (0/1)
- `awards_won`: Awards received (0/1)
- `avg_training_score`: Training score out of 100


## 4. Data Quality & Preprocessing

### 4.1 Missing Values Check

In [24]:
df.isna().sum()

employee_id                 0
department                  0
region                      0
education                 771
gender                      0
recruitment_channel         0
no_of_trainings             0
age                         0
previous_year_rating     1363
length_of_service           0
KPIs_met_more_than_80       0
awards_won                  0
avg_training_score          0
dtype: int64

In [25]:
# Checking percentage of missing values in each column
(df.isna().sum())/len(df)*100

employee_id              0.000000
department               0.000000
region                   0.000000
education                4.426710
gender                   0.000000
recruitment_channel      0.000000
no_of_trainings          0.000000
age                      0.000000
previous_year_rating     7.825688
length_of_service        0.000000
KPIs_met_more_than_80    0.000000
awards_won               0.000000
avg_training_score       0.000000
dtype: float64

In [26]:
# Filling missing categorical values with mode
df['education']=df['education'].fillna(df['education'].mode()[0])

In [27]:
# Filling missing numerical values with median
df['previous_year_rating']=df['previous_year_rating'].fillna(df['previous_year_rating'].median())

In [28]:
df.isna().sum()

employee_id              0
department               0
region                   0
education                0
gender                   0
recruitment_channel      0
no_of_trainings          0
age                      0
previous_year_rating     0
length_of_service        0
KPIs_met_more_than_80    0
awards_won               0
avg_training_score       0
dtype: int64

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17417 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17417 non-null  int64  
 1   department             17417 non-null  object 
 2   region                 17417 non-null  object 
 3   education              17417 non-null  object 
 4   gender                 17417 non-null  object 
 5   recruitment_channel    17417 non-null  object 
 6   no_of_trainings        17417 non-null  int64  
 7   age                    17417 non-null  int64  
 8   previous_year_rating   17417 non-null  float64
 9   length_of_service      17417 non-null  int64  
 10  KPIs_met_more_than_80  17417 non-null  int64  
 11  awards_won             17417 non-null  int64  
 12  avg_training_score     17417 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.7+ MB


### 4.2 Handling duplicates

In [31]:
df.duplicated().sum()

np.int64(2)

In [33]:
#The dataset contained 2 duplicate records
df=df.drop_duplicates()

In [34]:
df.duplicated().sum()

np.int64(0)

### 4.3 Data Types Fix

In [39]:
df.dtypes

employee_id                int64
department                object
region                    object
education                 object
gender                    object
recruitment_channel       object
no_of_trainings            int64
age                        int64
previous_year_rating     float64
length_of_service          int64
KPIs_met_more_than_80      int64
awards_won                 int64
avg_training_score         int64
dtype: object

In [40]:
# Converting rating to integer (was float due to previous NaN values)
df['previous_year_rating'] = df['previous_year_rating'].astype(int)

In [41]:
df.dtypes

employee_id               int64
department               object
region                   object
education                object
gender                   object
recruitment_channel      object
no_of_trainings           int64
age                       int64
previous_year_rating      int64
length_of_service         int64
KPIs_met_more_than_80     int64
awards_won                int64
avg_training_score        int64
dtype: object

---
### 4.1 Missing Values Handling
- `education` had ~4.4% missing → filled with mode
- `previous_year_rating` had ~7.8% missing → filled with median
- No missing values remain after cleaning

### 4.2 Duplicate Records
- Found **2 duplicate rows** → removed using drop_duplicates()
- Rechecked → 0 duplicates left

### 4.3 Data Type Fix
- `previous_year_rating` converted from float → int (1–5 scale)
- Other object columns kept as object (sufficient for EDA)

---

### ✔️ Status: Cleaning Complete
- No missing values
- No duplicates
- Corrected data type
- Ready for EDA



## 5: Exploratory Data Analysis (EDA)

### 5.1 Univariate Analysis

#### 5.1.1 Categorical Columns
Columns considered: department, education, recruitment_channel, gender, region

In [48]:
# Checking distribution of main categorical columns
print("Department: ",df['department'].value_counts().sort_values(ascending=False))
print("\nEducation: ",df['education'].value_counts().sort_values(ascending=False))
print("\nRecruitment Channel: ",df['recruitment_channel'].value_counts().sort_values(ascending=False))
print("\nGender: ",df['gender'].value_counts())


Department:  department
Sales & Marketing    5458
Operations           3524
Procurement          2240
Technology           2199
Analytics            1697
HR                    831
Finance               802
Legal                 332
R&D                   332
Name: count, dtype: int64

Education:  education
Bachelors          12288
Masters & above     4841
Below Secondary      286
Name: count, dtype: int64

Recruitment Channel:  recruitment_channel
other       9749
sourcing    7349
referred     317
Name: count, dtype: int64

Gender:  gender
m    12312
f     5103
Name: count, dtype: int64


### 5.1.1 Categorical Column Insights
- The department with the highest employee count is 5458
- The most common education level is Bachelors
- The primary recruitment channel is other channel
- Gender distribution shows Male is higher


### 5.1.2 Numerical Columns Summary
(Understanding key numerical features: age, avg_training_score, previous_year_rating, length_of_service, no_of_trainings)


In [52]:
# Summary of numerical columns
df.describe()

employee_id  no_of_trainings           age  previous_year_rating  \
count  17415.000000     17415.000000  17415.000000          17415.000000   
mean   39082.285214         1.250761     34.807982              3.318691   
std    22708.049139         0.595720      7.694464              1.218216   
min        3.000000         1.000000     20.000000              1.000000   
25%    19280.500000         1.000000     29.000000              3.000000   
50%    39115.000000         1.000000     33.000000              3.000000   
75%    58838.500000         1.000000     39.000000              4.000000   
max    78295.000000         9.000000     60.000000              5.000000   

       length_of_service  KPIs_met_more_than_80    awards_won  \
count       17415.000000           17415.000000  17415.000000   
mean            5.801493               0.358886      0.023371   
std             4.175632               0.479687      0.151082   
min             1.000000               0.000000      0.000000   
25%             3.000000               0.000000      0.000000   
50%             5.000000               0.000000      0.000000   
75%             7.000000               1.000000      0.000000   
max            34.000000               1.000000      1.000000   

       avg_training_score  
count        17415.000000  
mean            63.177720  
std             13.418314  
min             39.000000  
25%             51.000000  
50%             60.000000  
75%             75.000000  
max             99.000000

In [53]:
print("Age Summary:\n", df['age'].describe())
print("\nTraining Score Summary:\n", df['avg_training_score'].describe())
print("\nPrevious Year Rating Summary:\n", df['previous_year_rating'].describe())
print("\nYears of Service Summary:\n", df['length_of_service'].describe())
print("\nNumber of Trainings Summary:\n", df['no_of_trainings'].describe())

Age Summary:
 count    17415.000000
mean        34.807982
std          7.694464
min         20.000000
25%         29.000000
50%         33.000000
75%         39.000000
max         60.000000
Name: age, dtype: float64

Training Score Summary:
 count    17415.000000
mean        63.177720
std         13.418314
min         39.000000
25%         51.000000
50%         60.000000
75%         75.000000
max         99.000000
Name: avg_training_score, dtype: float64

Previous Year Rating Summary:
 count    17415.000000
mean         3.318691
std          1.218216
min          1.000000
25%          3.000000
50%          3.000000
75%          4.000000
max          5.000000
Name: previous_year_rating, dtype: float64

Years of Service Summary:
 count    17415.000000
mean         5.801493
std          4.175632
min          1.000000
25%          3.000000
50%          5.000000
75%          7.000000
max         34.000000
Name: length_of_service, dtype: float64

Number of Trainings Summary:
 count    17415.

In [ ]:
### 5.1.2 Numerical Column Insights

- **Employee Age:** The average age is ~35 years, with most employees between **29 to 39 years**, indicating a mid-career workforce.
- **Length of Service:** Median service length is **5 years**, suggesting employees typically stay around 3–7 years in the organization.
- **Training Count:** Most employees attend **1 training per year**, with very few taking more than 5 trainings.
- **Training Score:** The average training score is **63**, with most employees scoring between **51 to 75**, showing moderate training performance.
- **Performance Rating:** Previous year ratings mostly fall between **3 to 4**, meaning the overall performance level is average to good.
- **Awards:** Very few employees have won awards (**only ~2%**), indicating reward/recognition is limited.
- **KPI Achievement:** ~35% of employees achieved KPI scores above 80%, showing a limited high-performance group.
